api/lib/opentelemetry/trace/trace_id.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Trace
+    # A class that represents a span identifier. A valid span identifier is a
+    # 16-byte array with at least one non-zero byte.
+    class TraceId
+      INVALID_ID = 0
+      private_constant :INVALID_ID
+
+      MAX_128BIT_UNSIGNED_INT = (1 << 128) - 1
+      private_constant :MAX_128BIT_UNSIGNED_INT
+
+      def self.generate
+        id = rand(MAX_128BIT_UNSIGNED_INT)
+        return generate if id == INVALID_ID
+
+        new(id)
+      end
+
+      def initialize(id)
+        @id = id
+      end
+
+      # Checks if TraceId is valid
+      #
+      # A valid trace identifier is a 16-byte array with at least one non-zero
+      # byte.
+      #
+      # @return [Boolean] true if valid
+      def valid?
+        @id != INVALID_ID
+      end
+
+      def ==(other)
+        @id == other.id
+      end
+
+      protected
+
+      attr_reader :id
```

Comment:
Why does this need to be protected?

Reply:
In java implementation they implement the id as an integer but expose only the conversion methods (e.g. toLowerBase16): https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/SpanId.java . I currently did the same thing here. I used protected to be still able to compare two Ids in `#==`. If we expose the integer representation of the id then it probably could use a better name (e.g. to_i)